### PR TITLE
poplar: add initial poplar.mk

### PR DESCRIPTION
**Pull request by igoropaniuk**
Add initial poplar.mk, not integrated with buildroot yet.

We are using Poplar board intensively in GL, so why not to introduce the same setup the same for Hikey to avoid painful manual bring-up. 

Includes `flash-help` target with all steps about flashing partitions to the destination device.
Relates to: https://github.com/OP-TEE/manifest/pull/115/files

`Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>`

### CODE REVIEW

**Comment by vchong** [comment]
Would be nice to add `CFG_DRAM_SIZE_GB ?= 2` to `OPTEE_OS_COMMON_FLAGS`.

File: poplar.mk [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+UBOOT_BIN		?= $(UBOOT_PATH)/u-boot.bin
+OPTEE_BIN		?= $(OPTEE_PATH)/out/arm/core/tee-header_v2.bin
+OPTEE_BIN_EXTRA1	?= $(OPTEE_PATH)/out/arm/core/tee-pager_v2.bin
+OPTEE_BIN_EXTRA2	?= $(OPTEE_PATH)/out/arm/core/tee-pageable_v2.bin
+
+ROOTFS_BIN		?= linaro-stretch-developer-20170914-80.tar.gz
+ROOTFS_URL		?= http://download.djjproject.com/poplar/$(ROOTFS_BIN)
+
+PKG_OPTEE_VERSION	?= 1.0.0
```
Comment (vchong) [inline poplar.mk]: Why not `?= $(shell cd $(OPTEE_OS_PATH) && git describe)-0`?

File: poplar.mk [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+################################################################################
+# Toolchain
+################################################################################
+include toolchain.mk
+
+################################################################################
+# Folders
+################################################################################
+$(OUT_PATH):
```
Comment (vchong) [inline poplar.mk]: Who calls this target?

File: poplar.mk [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+	$(MAKE) -C $(OPTEE_TEST_PATH) \
+		CROSS_COMPILE_HOST="$(AARCH64_CROSS_COMPILE)" \
+		CROSS_COMPILE_TA="$(AARCH32_CROSS_COMPILE)" \
+		TA_DEV_KIT_DIR="$(OPTEE_PATH)/out/arm/export-ta_arm32" \
+		OPTEE_CLIENT_EXPORT="$(OPTEE_CLIENT_PATH)/out/export"
+
+.PHONY: optee-test-clean
+optee-test-clean:
+	$(MAKE) -C $(OPTEE_TEST_PATH) clean
+
```
Comment (vchong) [inline poplar.mk]: Can't you just depend on targets in `common.mk` for `optee-client`, `optee-client-clean`, `optee-test` and `optee-test-clean`?

File: poplar.mk [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+################################################################################
+# ARM Trusted Firmware
+################################################################################
+.PHONY: arm-tf
+arm-tf: u-boot optee-os
+	$(MAKE) -C $(ARM_TF_PATH) \
+		CROSS_COMPILE="$(AARCH64_CROSS_COMPILE)" \
```
Comment (vchong) [inline poplar.mk]: nit: `CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)" \`

File: poplar.mk [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+		all fip \
+		DEBUG=1 \
```
Comment (vchong) [inline poplar.mk]: `DEBUG=$(DEBUG)`

File: poplar.mk [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+LINUX_CLEANER_COMMON_FLAGS += ARCH=arm64
+
+linux-cleaner: linux-cleaner-common
+
+################################################################################
+# Build package
+################################################################################
+.PHONY: deb-package
+deb-package: optee-test optee-client
+	mkdir -p $(OUT_PATH)/debpkg
```
Comment (vchong) [inline poplar.mk]: Can remove

File: poplar.mk [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+	mkdir -p $(PKG_PATH)
```
Comment (vchong) [inline poplar.mk]: Can remove

File: poplar.mk [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+	@echo "4. Connect to Poplar board over serial console (run on host PC):"
+	@echo ""
+	@echo "   $$ screen /dev/ttyUSB0 115200"
+	@echo ""
+	@echo "5. Configure network interface in Poplar U-boot shell. If you can't"
+	@echo "get into U-boot console, press and hold Ctrl+C while booting:"
+	@echo ""
+	@echo "   => setenv ipaddr 192.168.0.2"
+	@echo "   => setenv netmask 255.255.255.0"
+	@echo "   => setenv serverip <your_host_ip_address>"
```
Comment (vchong) [inline poplar.mk]: Maybe helpful to mention that host ip is set to 192.168.0.3 as to example below.

File: poplar.mk [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+	@echo "   => tftp 0x08000000 recovery_files/install.scr"
+	@echo "   => source 0x08000000"
+	@echo ""
+	@echo "8. After successful flashing reboot your board (U-boot shell):"
+	@echo ""
+	@echo "   => reset"
+	@echo ""
+	@echo "9. Upload deb package using SCP (run on your host PC):"
+	@echo ""
+	@echo "   $$ scp optee_1.0.0.deb linaro@192.168.0.2:/tmp"
```
Comment (vchong) [inline poplar.mk]: `optee_${PKG_OPTEE_VERSION}.deb`

File: poplar.mk [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+
+LINUX_DTB		?= $(LINUX_PATH)/arch/arm64/boot/dts/hisilicon/hi3798cv200-poplar.dtb
+
+UBOOT_BIN		?= $(UBOOT_PATH)/u-boot.bin
+OPTEE_BIN		?= $(OPTEE_PATH)/out/arm/core/tee-header_v2.bin
+OPTEE_BIN_EXTRA1	?= $(OPTEE_PATH)/out/arm/core/tee-pager_v2.bin
+OPTEE_BIN_EXTRA2	?= $(OPTEE_PATH)/out/arm/core/tee-pageable_v2.bin
+
+ROOTFS_BIN		?= linaro-stretch-developer-20170914-80.tar.gz
+ROOTFS_URL		?= http://download.djjproject.com/poplar/$(ROOTFS_BIN)
```
[the remark this answers is not in the document]
Comment (igoropaniuk) [reply poplar.mk]: The problem with latest rootfs that is hosted on snapshots.linaro.org, is that it is flashed with errors to mmc (seems that poplar tools create wrong partitions images). That's why I started using the same image, that was mentioned in the official manual for bringing up debian on poplar (but, unfortunately, it's doesn't exist on snapshots.linaro.org anymore).

I don't remember exact error message, but will

File: poplar.mk (hunk 2 above)
Comment (igoropaniuk) [reply poplar.mk]: will delete it

File: poplar.mk [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+LLOADER_BIN		?= $(LLOADER_PATH)/l-loader.bin
+
+LINUX_DTB		?= $(LINUX_PATH)/arch/arm64/boot/dts/hisilicon/hi3798cv200-poplar.dtb
+
+UBOOT_BIN		?= $(UBOOT_PATH)/u-boot.bin
+OPTEE_BIN		?= $(OPTEE_PATH)/out/arm/core/tee-header_v2.bin
+OPTEE_BIN_EXTRA1	?= $(OPTEE_PATH)/out/arm/core/tee-pager_v2.bin
+OPTEE_BIN_EXTRA2	?= $(OPTEE_PATH)/out/arm/core/tee-pageable_v2.bin
+
+ROOTFS_BIN		?= linaro-stretch-developer-20170914-80.tar.gz
```
[the remark this answers is not in the document]
Comment (igoropaniuk) [reply poplar.mk]: @vchong we have to specify full filename, because it's fetched with wget, check `rootfs` make target.

**Comment by igoropaniuk** [comment]
@vchong @jbech-linaro the problem of integration with buildroot is that we depend on `poplar_recovery_builder.sh` script, which consumes rootfs in tar.gz format and then creates proper partition images and generates U-boot shell scripts, that simplify flashing of these partitions from U-boot shell.

As I understand, it will require some effort to create something similar on top of buildroot.

My suggestion is to merge what we have now (after addressing issues), and then proceed with buildroot integration in a separate PR.

File: poplar.mk (hunk 10 above)
[the remark this answers is not in the document]
Comment (igoropaniuk) [reply poplar.mk]: So the problem occurs when `unzip` is called for  recovery_files/partition3.2-of-8.gz.

```
ETH1: PHY(phyaddr=3, rgmii) link UP: DUPLEX=FULL : SPEED=100M
Using gmac1 device
TFTP from server 10.103.106.70; our IP address is 10.103.106.64
Filename 'recovery_files/partition3.2-of-8.gz'.
Load address: 0x8000000
Loading: #################################################################

done
Bytes transferred = 138279169 (83df901 hex)
Error: inflate() returned -3

MMC write: dev # 0, block # 794624, count 524288 ... 524288 blocks written: OK

```

I guess the root cause could be an insufficient amount of RAM space to store un-gziped partition (its gziped size is 132M )

```
out/recovery_files$ ls -lah
total 371M
drwxrwxrwx 2 igoropaniuk igoropaniuk 4,0K чер  5 15:52 .
drwxrwxr-x 4 igoropaniuk igoropaniuk 4,0K чер  5 15:33 ..
-rwxrwxrwx 1 igoropaniuk igoropaniuk 2,0M чер  5 15:31 fastboot.bin
-rwxrwxrwx 1 igoropaniuk igoropaniuk  645 чер  5 15:33 install
-rwxrwxrwx 1 igoropaniuk igoropaniuk  169 чер  5 15:31 install-layout
-rwxrwxrwx 1 igoropaniuk igoropaniuk  241 чер  5 15:31 install-layout.scr
-rwxrwxrwx 1 igoropaniuk igoropaniuk  168 чер  5 15:32 install-partition1
-rwxrwxrwx 1 igoropaniuk igoropaniuk  240 чер  5 15:32 install-partition1.scr
-rwxrwxrwx 1 igoropaniuk igoropaniuk  167 чер  5 15:32 install-partition2
-rwxrwxrwx 1 igoropaniuk igoropaniuk  239 чер  5 15:32 install-partition2.scr
-rwxrwxrwx 1 igoropaniuk igoropaniuk 1,2K чер  5 15:33 install-partition3
-rwxrwxrwx 1 igoropaniuk igoropaniuk 1,3K чер  5 15:33 install-partition3.scr
-rwxrwxrwx 1 igoropaniuk igoropaniuk  717 чер  5 15:33 install.scr
-rwxrwxrwx 1 igoropaniuk igoropaniuk   86 чер  5 15:31 mbr.bin.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 396K чер  5 15:32 partition1.1-of-1.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 6,1M чер  5 15:32 partition2.1-of-1.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk  42M чер  5 15:32 partition3.1-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 132M чер  5 15:32 partition3.2-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk  93M чер  5 15:32 partition3.3-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk  77M чер  5 15:33 partition3.4-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk  20M чер  5 15:33 partition3.5-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 255K чер  5 15:33 partition3.6-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 255K чер  5 15:33 partition3.7-of-8.gz
-rwxrwxrwx 1 igoropaniuk igoropaniuk 161K чер  5 15:33 partition3.8-of-8.gz

```

**Comment by jbech-linaro** [comment]
> @vchong @jbech-linaro the problem of integration with buildroot is that we depend on poplar_recovery_builder.sh
> ...
> My suggestion is to merge what we have now (after addressing issues), and then proceed with buildroot integration in a separate PR.

That's fine, there no strict rule that we shall run Buildroot on everything either. It's certainly good if we can do that and something we strive for, since then we have a common way of doing things. But right now, this is new stuff, so I'm fine with your suggestions here.

File: poplar.mk (hunk 10 above)
[the remark this answers is not in the document]
Comment (igoropaniuk) [reply poplar.mk]:
```
$ git diff
diff --git a/poplar_recovery_builder.sh b/poplar_recovery_builder.sh
index e4493b4..c9dd0e2 100755
--- a/poplar_recovery_builder.sh
+++ b/poplar_recovery_builder.sh
@@ -18,7 +18,7 @@ EMMC_SIZE=14942208    # 7296 MB in sectors (not hex)
 CHUNK_SIZE=524288      # Partition image chuck size in sectors (not hex)
 
 IN_ADDR=0x08000000     # Buffer address for compressed data in from USB (hex)
-OUT_ADDR=0x10000000    # Buffer address for uncompressed data for MMC (hex)
+OUT_ADDR=0x11000000    # Buffer address for uncompressed data for MMC (hex)
 SUB_ADDR=0x07800000    # Buffer address for sub-installer scripts
 
 EMMC_DEV=/dev/mmcblk0  # Linux path to main eMMC device on target
@@ -118,7 +118,7 @@ function parseargs() {
        PARTS=$1
        shift
 
-       LOAD_COMMAND=tftp
+       LOAD_COMMAND=tftpboot
        INPUT_FILES=""
 
        while [ $# -gt 0 ]; do

```

[the remark this answers is not in the document]
Comment (igoropaniuk) [reply poplar.mk]: https://github.com/96boards-poplar/poplar-tools/pull/4

**Comment by igoropaniuk** [comment]
All comments addressed

File: poplar.mk [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+
+UBOOT_BIN		?= $(UBOOT_PATH)/u-boot.bin
+OPTEE_BIN		?= $(OPTEE_PATH)/out/arm/core/tee-header_v2.bin
+OPTEE_BIN_EXTRA1	?= $(OPTEE_PATH)/out/arm/core/tee-pager_v2.bin
+OPTEE_BIN_EXTRA2	?= $(OPTEE_PATH)/out/arm/core/tee-pageable_v2.bin
+
+ROOTFS_BIN		?= linaro-stretch-developer-*.tar.gz
+ROOTFS_URL		?= http://snapshots.linaro.org/debian/images/stretch/developer-arm64/latest/$(ROOTFS_BIN)
+
+PKG_OPTEE_VERSION	?= $(shell cd $(OPTEE_OS_PATH) && git describe)
```
Comment (vchong) [inline poplar.mk]: Need to end with `-0`. See https://github.com/OP-TEE/build/pull/154#discussion_r121323471 for details.

Other than that, `Reviewed-by: Victor Chong <victor.chong@linaro.org>`

Comment (igoropaniuk) [reply poplar.mk]: fixed, thanks!

**Comment by igoropaniuk** [comment]
tag applied, rebased and ready for merge

**Comment by igoropaniuk** [comment]
BTW, https://github.com/96boards-poplar/poplar-tools/pull/4 was also merged

**Pull request merged**